### ISSUE TITLE
500 after updating to latest version

### ISSUE THREAD

**Issue by FiloSottile** [ndarilek/dokuwiki-sandstorm#5]
```
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns
waiting for php5-fpm to be available at /var/run/php5-fpm.sock
2016/10/27 14:55:17 [error] 670#0: *2 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Cannot redeclare class DokuWiki_Action_Plugin in /var/lib/dokuwiki/lib/plugins/action.php on line 0" while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "GET / HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: 
```

**Comment by FiloSottile** [comment]
Fixed manually by deleting `lib/dokuwiki/lib/plugins/action.php` from a backup and restoring.

**Comment by ndarilek** [comment]
Hmm, sorry, not sure what to do about this one. I'm not a PHP developer 
and am in a bit over my head.

Can you take a backup of the grain and compare the referenced file to 
lib/plugins/action.php as shipped in DokuWiki? Maybe the rsync isn't 
updating this as it should. If they don't match, try replacing the file 
in the backup and rezipping that? I'll see about shipping a new version 
that fixes things if the files are indeed out of sync.

Thanks for testing this. I'm reading your issues but am a bit low on 
time these days.

**Comment by ndarilek** [comment]
Hmm, not sure how to fix this in the upgrade case. I'm rsyncing in the 
plugins directory to update the files Doku provides, but I don't want to 
delete missing files since someone may have their own plugins installed. 
This is a case where the fact that Doku doesn't separate its own files 
from user-installed files is going to bite us.

Open to thoughts if anyone reading this has them.

**Comment by ndarilek** [comment]
Do you happen to have the original version of this file that failed for 
you? Looking at the launcher script, it should have rsynced in the newer 
version, and this file wasn't removed in any recent updates. In short, I 
think I've coded for this case but it seems like that may not be working.

Thanks.

**Comment by kentonv** [comment]
Another user reported this problem on IRC. Is this affecting everyone who upgraded? I think it would be a good idea to update the package again such that it manually deletes this specific file at startup, as a hack to fix this problem.

**Comment by ndarilek** [comment]
I don't know if deleting the file _is_ the correct solution, though. 
When I look in the latest source, it is still present.

What would be helpful is someone having an older grain backup they'd not 
mind sharing with me. I'd like to compare the old file with the newer 
version and determine if I have my rsync incantation wrong.

Failing that, I think I have older SPKs around, and can stage an upgrade 
of my own. Fixing this is a priority for me since we're going to try 
another switch to Sandstorm with standalone grains, and I need Doku working.

Thanks.

**Comment by ndarilek** [comment]
Unfortunately I'm unable to duplicate this by:
1. Uploading the previous version's SPK to my vagrant-spk Sandstorm instance.
2. Creating a wiki.
3. Uploading the current version.
4. Clicking the Upgrade button to upgrade my grain.
5. Clicking the link to the open grain.
6. Clicking Restart.

So I won't be able to fix this unless someone sends me a copy of the broken PHP file in question so I can figure out which version is in the grain. Or better yet, if someone can send me a backup of a broken grain, then I can use that. But this error shouldn't be happening.

I have a semi-production wiki that I'll try duplicating this on. Otherwise I'm not sure where to go next. Sorry.

**Comment by funwhilelost** [comment]
[Zine Wiki.zip](https://github.com/ndarilek/dokuwiki-sandstorm/files/637968/Zine.Wiki.zip)

That should help out if you want to triage a broken, older grain.

Thanks for your help.

**Comment by VK3FNG** [comment]
Just chiming in - I have this problem also. Deleting action.php did not work for me.

I've got quite a large wiki of personal information here, and I'm a little worried I'm not going to be able to get it back.

`** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns
waiting for php5-fpm to be available at /var/run/php5-fpm.sock
2017/01/26 02:24:39 [error] 673#0: *2 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Cannot redeclare class DokuWiki_Action_Plugin in /var/lib/dokuwiki/lib/plugins/action.php on line 0" while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "GET / HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "d9eda1042439d08c3d7451fe8b09b81f.oasis.sandstorm.io"
** SANDSTORM SUPERVISOR: Grain shutdown requested.
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns
waiting for php5-fpm to be available at /var/run/php5-fpm.sock
waiting for php5-fpm to be available at /var/run/php5-fpm.sock
2017/01/26 02:28:03 [error] 673#0: *2 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Cannot redeclare class DokuWiki_Action_Plugin in /var/lib/dokuwiki/lib/plugins/action.php on line 0" while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "GET / HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "7ba4cd76377358258823f88831470bbd.oasis.sandstorm.io"
`

**Comment by funwhilelost** [comment]
@VK3FNG No worries about getting your content back.  The cool thing about DokuWiki is that all the content you've added is just sitting there in text files.

If you download a backup of the grain you can open the Zip file and find the text files inside `data/lib/dokuwiki/data/pages`.  If you start a new DokuWiki grain, download a backup of that grain, open the zip, add your text files from the old grain, then zip up and "restore" the grain you should be able to restore everything into the new version.

**Comment by ndarilek** [comment]
Thanks. I'm going to try looking at this tomorrow. Honestly I really 
don't like how Doku is laid out, and its method of intermingling core 
plugins, user-installed plugins and content seems like a bit of a 
dumpster fire. I just wanted to capture that in a hopefully fool-proof 
way, but in the process only discovered that I'm a bigger fool than I 
thought. :) I don't know why this would fail, as the intent was to copy 
up-to-date core files into the sandbox.

**Comment by funwhilelost** [comment]
Much appreciated, @ndarilek

**Comment by ndarilek** [comment]
Apologies for taking forever to get to this. Honestly I can't stand administering DokuWiki. :/

The issue appears to be that, somehow, a dangling self-recursive symbolic link landed in some of your /var directories. So /var/lib/dokuwiki/lib/plugins/plugins was a link to /var/lib/dokuwiki/lib/plugins, which caused the redefinition. This appeared to happen in a few other places, too. Strangely I never saw this in any of *my* upgrades, and I can't duplicate it locally other than by starting from a broken backup.

I just pushed a new version which a) shouldn't exhibit this problem and b) should clean up old grains where it existed. If anyone is still using this package, it should be available on the experimental app market (not sure immediately how to get to that, maybe add ?experimental=1 to https://apps.sandstorm.io?) I tested this on the broken wiki backup from this issue and it appeared to work.

I'm declaring this closed since, if it is still happening, either my fix didn't work or we've got a different problem than I believe.
